src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/service/VisitServiceTest.kt

```diff
@@ -0,0 +1,181 @@
+import java.time.LocalDateTime
+import java.time.LocalTime
+import java.util.Optional
+
+private const val offenderBookingId = -9L
+private const val visitId = -8L
+private const val offenderNo = "A1234AA"
+private const val prisonId = "SWI"
+
+internal class VisitServiceTest {
+
+  private val visitRepository: VisitRepository = mock()
+  private val visitVisitorRepository: VisitVisitorRepository = mock()
+  private val offenderBookingRepository: OffenderBookingRepository = mock()
+  private val offenderVisitBalanceRepository: OffenderVisitBalanceRepository = mock()
+  private val offenderVisitBalanceAdjustmentRepository: OffenderVisitBalanceAdjustmentRepository = mock()
+  private val eventStatusRepository: ReferenceCodeRepository<EventStatus> = mock()
+  private val visitTypeRepository: ReferenceCodeRepository<VisitType> = mock()
+  private val visitStatusRepository: ReferenceCodeRepository<VisitStatus> = mock()
+  private val agencyRepository: AgencyLocationRepository = mock()
+  private val telemetryClient: TelemetryClient = mock()
+
+  private val visitService: VisitService = VisitService(
+    visitRepository,
+    visitVisitorRepository,
+    offenderBookingRepository,
+    offenderVisitBalanceRepository,
+    offenderVisitBalanceAdjustmentRepository,
+    eventStatusRepository,
+    visitTypeRepository,
+    visitStatusRepository,
+    agencyRepository,
+    telemetryClient,
+  )
+
+  val visitType = VisitType("SCON", "desc")
+  val visitStatus = VisitStatus("SCH", "desc")
+  val eventStatus = EventStatus("SCH", "desc")
+
+  val createVisitRequest = CreateVisitRequest(
+    visitType = "SCON",
+    offenderNo = offenderNo,
+    startTime = LocalDateTime.of(2021, 11, 4, 12, 5),
+    endTime = LocalTime.of(13, 4),
+    prisonId = prisonId,
+    visitorPersonIds = listOf(45L, 46L),
+    decrementBalances = true,
+  )
+
+  @BeforeEach
+  fun setup() {
+    whenever(offenderBookingRepository.findByOffenderNomsIdAndActive(offenderNo, true)).thenReturn(
+      Optional.of(
+        OffenderBooking(
+          bookingId = offenderBookingId,
+          bookingBeginDate = LocalDateTime.now(),
+        )
+      )
+    )
+    whenever(visitTypeRepository.findById(VisitType.pk("SCON"))).thenReturn(Optional.of(visitType))
+    whenever(visitStatusRepository.findById(VisitStatus.pk("SCH"))).thenReturn(Optional.of(visitStatus))
+    whenever(eventStatusRepository.findById(EventStatus.SCHEDULED_APPROVED)).thenReturn(Optional.of(eventStatus))
+    whenever(agencyRepository.findById(prisonId)).thenReturn(Optional.of(AgencyLocation(prisonId, "desc")))
+
+    whenever(offenderVisitBalanceRepository.findById(offenderBookingId)).thenReturn(
+      Optional.of(
+        OffenderVisitBalance(
+          offenderBookingId = offenderBookingId,
+          remainingVisitOrders = 3,
+          remainingPrivilegedVisitOrders = 5,
+        )
+      )
+    )
+  }
+
+  @Nested
+  internal inner class createVisit {
+    @Test
+    fun success() {
+
```

Comment:
maybe break this test in to smaller ones since it is quite hard to follow what is important and why things happen?

fun `will return the visit id after creating`()

fun `will decrement the balance`()

fun `will set the core visit data`()

etc etc